lib/SILOptimizer/Analysis/ColdBlockInfo.cpp

```diff
@@ -221,31 +238,99 @@ bool ColdBlockInfo::inferFromEdgeProfile(SILBasicBlock *BB) {
   SmallVector<ProfileCounter, 2> succCount;
 
   // Current analysis only accurately handles blocks with 2 successors,
-  // especially since we only have two temperatures.
+  // especially since we only have two temperatures (cold/warm).
+  // TODO: With propagation algorithm, this limitation can be removed to support
+  // multi-way branches (switch statements) and arbitrary successor counts.
   if (BB->getNumSuccessors() != 2)
     return false;
 
-  // Check the successor edges for profile data.
+  // First pass: collect all counters and check for evidence of profiling
+  // Note: std::optional<ProfileCounter> is used to distinguish between:
+  //   - std::nullopt: missing profile data (profiler didn't record this edge)
+  //   - ProfileCounter(0): definitely zero executions (profiler recorded 0)
+  // This distinction is crucial for both the current optimistic strategy and
+  // future propagation algorithms, which need to know where data is missing.
+  bool hasAnyNonZeroCount = false;
+  bool hasAnyMissingData = false;
+  SmallVector<std::optional<ProfileCounter>, 2> counters;
+
   for (auto const &succ : BB->getSuccessors()) {
     auto counter = succ.getCount();
 
-    // Can't make an inference if there's profile data missing for a successor.
-    // FIXME: there are techniques to determine a missing count;
-    //        see the SamplePGO paper by Diego Novillo.
-    if (!counter)
-      return false;
+    if (!counter.hasValue()) {
+      hasAnyMissingData = true;
+      counters.push_back(std::nullopt);
+    } else {
+      counters.push_back(counter);
+      if (counter.getValue() > 0) {
+        hasAnyNonZeroCount = true;
+      }
+    }
+  }
 
-    succCount.push_back(counter);
+  // Apply the optimistic strategy for missing data:
+  // If we have evidence of profiling (at least one non-zero count),
+  // treat missing data as zero. Otherwise, bail out conservatively.
+  // This prevents marking unprofiled code as entirely cold while enabling
+  // optimizations when profile data is partial.
+  if (hasAnyMissingData && !hasAnyNonZeroCount) {
+    // No evidence of profiling - be conservative
+    LLVM_DEBUG(llvm::dbgs() << "ColdBlockInfo: no evidence of profiling for "
+                            << toString(BB) << " (conservative approach)\n");
+    return false;
+  }
 
-    auto didSaturate = totalCount.add_saturating(counter);
+  // Second pass: build the count vector, treating missing as zero if optimistic
+  for (size_t i = 0; i < counters.size(); i++) {
+    ProfileCounter count;
+    if (counters[i].has_value()) {
+      count = counters[i].value();
+    } else {
+      // We're being optimistic - treat missing as zero
+      count = ProfileCounter(0);
+      LLVM_DEBUG(llvm::dbgs()
+        << "ColdBlockInfo: treating missing profile data as zero for "
+        << toString(BB->getSuccessors()[i])
+        << " (optimistic strategy - found non-zero counts on other edges)\n");
+    }
```

Comment:
This is applying the optimistic strategy unconditionally. I would expect it to be aligned with the flag.